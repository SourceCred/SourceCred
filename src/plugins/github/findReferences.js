// @flow

function findAllMatches(re: RegExp, s: string): any[] {
  // modified from: https://stackoverflow.com/a/6323598
  let m;
  const matches = [];
  do {
    m = re.exec(s);
    if (m) {
      matches.push(m);
    }
  } while (m);
  return matches;
}

export function findReferences(body: string): string[] {
  // Note to maintainer: If it becomes necessary to encode references in a
  // richer format, consider implementing the type signature described in
  // https://github.com/sourcecred/sourcecred/pull/130#pullrequestreview-113849998
  return [
    ...findNumericReferences(body),
    ...findGithubUrlReferences(body),
    ...findUsernameReferences(body),
  ];
}

function findNumericReferences(body: string): string[] {
  return findAllMatches(/(?:\W|^)(#\d+)(?:\W|$)/g, body).map((x) => x[1]);
}

function findUsernameReferences(body: string): string[] {
  return findAllMatches(/(?:\W|^)(@[a-zA-Z0-9-]+)(?:\W|$)/g, body).map(
    (x) => x[1]
  );
}

function findGithubUrlReferences(body: string): string[] {
  const githubNamePart = /([a-zA-Z0-9_-]+)/.source;
  const urlRegex = new RegExp(
    "" +
      /(?:\W|^)http(?:s)?:\/\/github.com\//.source +
      githubNamePart +
      /\//.source +
      githubNamePart +
      /\/(issues|pull)\//.source +
      /(\d+)/.source +
      /(#(issue|issuecomment|pullrequestreview|discussion_r)-?(\d+))?/.source +
      /(?:\W|$)/.source,
    "gm"
  );
  return findAllMatches(urlRegex, body).map((match) => match[0].trim());
}
