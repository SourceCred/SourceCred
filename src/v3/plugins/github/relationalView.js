// @flow

import * as N from "./nodes";
import * as Q from "./graphql";
import * as GitNode from "../git/nodes";

import {
  reviewUrlToId,
  issueCommentUrlToId,
  pullCommentUrlToId,
  reviewCommentUrlToId,
} from "./urlIdParse";

export type RepoEntry = {|
  +type: typeof N.REPO_TYPE,
  +address: N.RepoAddress,
  +url: string,
  +issues: IssueEntry[],
  +pulls: PullEntry[],
|};

export type IssueEntry = {|
  +type: typeof N.ISSUE_TYPE,
  +address: N.IssueAddress,
  +title: string,
  +body: string,
  +url: string,
  +comments: CommentEntry[],
  +nominalAuthor: ?UserlikeEntry,
|};

export type PullEntry = {|
  +type: typeof N.PULL_TYPE,
  +address: N.PullAddress,
  +title: string,
  +body: string,
  +url: string,
  +comments: CommentEntry[],
  +reviews: ReviewEntry[],
  +mergedAs: ?GitNode.CommitAddress,
  +nominalAuthor: ?UserlikeEntry,
|};

export type ReviewEntry = {|
  +type: typeof N.REVIEW_TYPE,
  +address: N.ReviewAddress,
  +body: string,
  +url: string,
  +comments: CommentEntry[],
  +state: Q.ReviewState,
  +nominalAuthor: ?UserlikeEntry,
|};

export type CommentEntry = {|
  +type: typeof N.COMMENT_TYPE,
  +address: N.CommentAddress,
  +body: string,
  +url: string,
  +nominalAuthor: ?UserlikeEntry,
|};

export type UserlikeEntry = {|
  +type: typeof N.USERLIKE_TYPE,
  +address: N.UserlikeAddress,
  +url: string,
|};

export type Entry =
  | RepoEntry
  | IssueEntry
  | PullEntry
  | ReviewEntry
  | CommentEntry
  | UserlikeEntry;

export class RelationalView {
  _repos: Map<N.RawAddress, RepoEntry>;
  _issues: Map<N.RawAddress, IssueEntry>;
  _pulls: Map<N.RawAddress, PullEntry>;
  _comments: Map<N.RawAddress, CommentEntry>;
  _reviews: Map<N.RawAddress, ReviewEntry>;
  _userlikes: Map<N.RawAddress, UserlikeEntry>;

  constructor(data: Q.GithubResponseJSON) {
    this._repos = new Map();
    this._issues = new Map();
    this._pulls = new Map();
    this._comments = new Map();
    this._reviews = new Map();
    this._userlikes = new Map();
    this._addRepo(data.repository);
  }

  repos(): Iterator<RepoEntry> {
    return this._repos.values();
  }
  repo(address: N.RepoAddress): ?RepoEntry {
    return this._repos.get(N.toRaw(address));
  }

  issues(): Iterator<IssueEntry> {
    return this._issues.values();
  }
  issue(address: N.IssueAddress): ?IssueEntry {
    return this._issues.get(N.toRaw(address));
  }

  pulls(): Iterator<PullEntry> {
    return this._pulls.values();
  }
  pull(address: N.PullAddress): ?PullEntry {
    return this._pulls.get(N.toRaw(address));
  }

  comments(): Iterator<CommentEntry> {
    return this._comments.values();
  }
  comment(address: N.CommentAddress): ?CommentEntry {
    return this._comments.get(N.toRaw(address));
  }

  reviews(): Iterator<ReviewEntry> {
    return this._reviews.values();
  }
  review(address: N.ReviewAddress): ?ReviewEntry {
    return this._reviews.get(N.toRaw(address));
  }

  userlikes(): Iterator<UserlikeEntry> {
    return this._userlikes.values();
  }
  userlike(address: N.UserlikeAddress): ?UserlikeEntry {
    return this._userlikes.get(N.toRaw(address));
  }

  _addRepo(json: Q.RepositoryJSON) {
    const address: N.RepoAddress = {
      type: N.REPO_TYPE,
      owner: json.owner.login,
      name: json.name,
    };
    const entry: RepoEntry = {
      address,
      url: json.url,
      issues: json.issues.nodes.map((x) => this._addIssue(address, x)),
      pulls: json.pulls.nodes.map((x) => this._addPull(address, x)),
      type: N.REPO_TYPE,
    };
    const raw = N.toRaw(address);
    this._repos.set(raw, entry);
  }

  _addIssue(repo: N.RepoAddress, json: Q.IssueJSON): IssueEntry {
    const address: N.IssueAddress = {
      type: N.ISSUE_TYPE,
      number: String(json.number),
      repo,
    };
    const entry: IssueEntry = {
      address,
      url: json.url,
      comments: json.comments.nodes.map((x) => this._addComment(address, x)),
      nominalAuthor: this._addNullableAuthor(json.author),
      body: json.body,
      title: json.title,
      type: N.ISSUE_TYPE,
    };
    this._issues.set(N.toRaw(address), entry);
    return entry;
  }

  _addPull(repo: N.RepoAddress, json: Q.PullJSON): PullEntry {
    const address: N.PullAddress = {
      type: N.PULL_TYPE,
      number: String(json.number),
      repo,
    };
    const mergedAs =
      json.mergeCommit == null
        ? null
        : {
            type: GitNode.COMMIT_TYPE,
            hash: json.mergeCommit.oid,
          };

    const entry: PullEntry = {
      address,
      url: json.url,
      comments: json.comments.nodes.map((x) => this._addComment(address, x)),
      reviews: json.reviews.nodes.map((x) => this._addReview(address, x)),
      nominalAuthor: this._addNullableAuthor(json.author),
      body: json.body,
      title: json.title,
      mergedAs,
      type: N.PULL_TYPE,
    };
    this._pulls.set(N.toRaw(address), entry);
    return entry;
  }

  _addReview(pull: N.PullAddress, json: Q.ReviewJSON): ReviewEntry {
    const address: N.ReviewAddress = {
      type: N.REVIEW_TYPE,
      id: reviewUrlToId(json.url),
      pull,
    };
    const entry: ReviewEntry = {
      address,
      url: json.url,
      state: json.state,
      comments: json.comments.nodes.map((x) => this._addComment(address, x)),
      body: json.body,
      nominalAuthor: this._addNullableAuthor(json.author),
      type: N.REVIEW_TYPE,
    };
    this._reviews.set(N.toRaw(address), entry);
    return entry;
  }

  _addComment(
    parent: N.IssueAddress | N.PullAddress | N.ReviewAddress,
    json: Q.CommentJSON
  ): CommentEntry {
    const id = (function() {
      switch (parent.type) {
        case N.ISSUE_TYPE:
          return issueCommentUrlToId(json.url);
        case N.PULL_TYPE:
          return pullCommentUrlToId(json.url);
        case N.REVIEW_TYPE:
          return reviewCommentUrlToId(json.url);
        default:
          // eslint-disable-next-line no-unused-expressions
          (parent.type: empty);
          throw new Error(`Unexpected comment parent type: ${parent.type}`);
      }
    })();
    const address: N.CommentAddress = {type: N.COMMENT_TYPE, id, parent};
    const entry: CommentEntry = {
      address,
      url: json.url,
      nominalAuthor: this._addNullableAuthor(json.author),
      body: json.body,
      type: N.COMMENT_TYPE,
    };
    this._comments.set(N.toRaw(address), entry);
    return entry;
  }

  _addNullableAuthor(json: Q.NullableAuthorJSON): ?UserlikeEntry {
    if (json == null) {
      return null;
    } else {
      const address: N.UserlikeAddress = {
        type: N.USERLIKE_TYPE,
        login: json.login,
      };
      const entry: UserlikeEntry = {
        address,
        url: json.url,
        type: N.USERLIKE_TYPE,
      };
      this._userlikes.set(N.toRaw(address), entry);
      return entry;
    }
  }
}
