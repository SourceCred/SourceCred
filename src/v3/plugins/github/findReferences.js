// @flow

import * as N from "./nodes";
import * as R from "./relationalView";
import {parseReferences} from "./parseReferences";

export type GithubReference = {|
  src: N.TextContentAddress,
  dst: N.ReferentAddress,
|};

type TextContentEntry =
  | R.IssueEntry
  | R.PullEntry
  | R.CommentEntry
  | R.ReviewEntry;
export function* findReferences(
  view: R.RelationalView
): Iterator<GithubReference> {
  function* allGithubEntities(): Iterator<R.Entry> {
    yield* view.repos();
    yield* view.issues();
    yield* view.pulls();
    yield* view.reviews();
    yield* view.comments();
    yield* view.userlikes();
  }

  const refToAddress: Map<string, N.StructuredAddress> = new Map();
  for (const e of allGithubEntities()) {
    const a = e.address;
    refToAddress.set(e.url, a);
    switch (e.type) {
      case "USERLIKE":
        refToAddress.set(`@${e.address.login}`, a);
        break;
      case "ISSUE":
        refToAddress.set(`#${e.address.number}`, a);
        break;
      case "PULL":
        refToAddress.set(`#${e.address.number}`, a);
        break;
      default:
        break;
    }
  }

  function* allTextContentEntries(): Iterator<TextContentEntry> {
    yield* view.issues();
    yield* view.pulls();
    yield* view.reviews();
    yield* view.comments();
  }

  for (const e of allTextContentEntries()) {
    for (const ref of parseReferences(e.body)) {
      const refAddress = refToAddress.get(ref);
      if (refAddress != null) {
        yield {src: e.address, dst: refAddress};
      }
    }
  }
}
