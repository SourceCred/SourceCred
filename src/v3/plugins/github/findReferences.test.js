// @flow

import {findReferences} from "./findReferences.js";

import * as R from "./relationalView";

describe("plugins/github/findReferences", () => {
  it("matches snapshot on exmaple-github data", () => {
    const data = require("./demoData/example-github");
    const view = new R.RelationalView(data);
    const references = Array.from(findReferences(view));
    expect(references).toMatchSnapshot();
  });
});
