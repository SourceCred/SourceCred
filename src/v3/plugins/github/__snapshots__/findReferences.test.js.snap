// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`plugins/github/findReferences matches snapshot on exmaple-github data 1`] = `
Array [
  Object {
    "dst": Object {
      "number": "1",
      "repo": Object {
        "name": "example-github",
        "owner": "sourcecred",
        "type": "REPO",
      },
      "type": "ISSUE",
    },
    "src": Object {
      "number": "2",
      "repo": Object {
        "name": "example-github",
        "owner": "sourcecred",
        "type": "REPO",
      },
      "type": "ISSUE",
    },
  },
  Object {
    "dst": Object {
      "login": "wchargin",
      "type": "USERLIKE",
    },
    "src": Object {
      "number": "5",
      "repo": Object {
        "name": "example-github",
        "owner": "sourcecred",
        "type": "REPO",
      },
      "type": "PULL",
    },
  },
  Object {
    "dst": Object {
      "login": "wchargin",
      "type": "USERLIKE",
    },
    "src": Object {
      "number": "9",
      "repo": Object {
        "name": "example-github",
        "owner": "sourcecred",
        "type": "REPO",
      },
      "type": "PULL",
    },
  },
  Object {
    "dst": Object {
      "number": "6",
      "repo": Object {
        "name": "example-github",
        "owner": "sourcecred",
        "type": "REPO",
      },
      "type": "ISSUE",
    },
    "src": Object {
      "id": "373768703",
      "parent": Object {
        "number": "2",
        "repo": Object {
          "name": "example-github",
          "owner": "sourcecred",
          "type": "REPO",
        },
        "type": "ISSUE",
      },
      "type": "COMMENT",
    },
  },
  Object {
    "dst": Object {
      "id": "373768538",
      "parent": Object {
        "number": "6",
        "repo": Object {
          "name": "example-github",
          "owner": "sourcecred",
          "type": "REPO",
        },
        "type": "ISSUE",
      },
      "type": "COMMENT",
    },
    "src": Object {
      "id": "373768850",
      "parent": Object {
        "number": "2",
        "repo": Object {
          "name": "example-github",
          "owner": "sourcecred",
          "type": "REPO",
        },
        "type": "ISSUE",
      },
      "type": "COMMENT",
    },
  },
  Object {
    "dst": Object {
      "number": "5",
      "repo": Object {
        "name": "example-github",
        "owner": "sourcecred",
        "type": "REPO",
      },
      "type": "PULL",
    },
    "src": Object {
      "id": "385576185",
      "parent": Object {
        "number": "2",
        "repo": Object {
          "name": "example-github",
          "owner": "sourcecred",
          "type": "REPO",
        },
        "type": "ISSUE",
      },
      "type": "COMMENT",
    },
  },
  Object {
    "dst": Object {
      "id": "100313899",
      "pull": Object {
        "number": "5",
        "repo": Object {
          "name": "example-github",
          "owner": "sourcecred",
          "type": "REPO",
        },
        "type": "PULL",
      },
      "type": "REVIEW",
    },
    "src": Object {
      "id": "385576220",
      "parent": Object {
        "number": "2",
        "repo": Object {
          "name": "example-github",
          "owner": "sourcecred",
          "type": "REPO",
        },
        "type": "ISSUE",
      },
      "type": "COMMENT",
    },
  },
  Object {
    "dst": Object {
      "id": "171460198",
      "parent": Object {
        "id": "100313899",
        "pull": Object {
          "number": "5",
          "repo": Object {
            "name": "example-github",
            "owner": "sourcecred",
            "type": "REPO",
          },
          "type": "PULL",
        },
        "type": "REVIEW",
      },
      "type": "COMMENT",
    },
    "src": Object {
      "id": "385576248",
      "parent": Object {
        "number": "2",
        "repo": Object {
          "name": "example-github",
          "owner": "sourcecred",
          "type": "REPO",
        },
        "type": "ISSUE",
      },
      "type": "COMMENT",
    },
  },
  Object {
    "dst": Object {
      "login": "wchargin",
      "type": "USERLIKE",
    },
    "src": Object {
      "id": "385576273",
      "parent": Object {
        "number": "2",
        "repo": Object {
          "name": "example-github",
          "owner": "sourcecred",
          "type": "REPO",
        },
        "type": "ISSUE",
      },
      "type": "COMMENT",
    },
  },
  Object {
    "dst": Object {
      "number": "1",
      "repo": Object {
        "name": "example-github",
        "owner": "sourcecred",
        "type": "REPO",
      },
      "type": "ISSUE",
    },
    "src": Object {
      "id": "385576920",
      "parent": Object {
        "number": "2",
        "repo": Object {
          "name": "example-github",
          "owner": "sourcecred",
          "type": "REPO",
        },
        "type": "ISSUE",
      },
      "type": "COMMENT",
    },
  },
  Object {
    "dst": Object {
      "number": "2",
      "repo": Object {
        "name": "example-github",
        "owner": "sourcecred",
        "type": "REPO",
      },
      "type": "ISSUE",
    },
    "src": Object {
      "id": "385576920",
      "parent": Object {
        "number": "2",
        "repo": Object {
          "name": "example-github",
          "owner": "sourcecred",
          "type": "REPO",
        },
        "type": "ISSUE",
      },
      "type": "COMMENT",
    },
  },
  Object {
    "dst": Object {
      "number": "3",
      "repo": Object {
        "name": "example-github",
        "owner": "sourcecred",
        "type": "REPO",
      },
      "type": "PULL",
    },
    "src": Object {
      "id": "385576920",
      "parent": Object {
        "number": "2",
        "repo": Object {
          "name": "example-github",
          "owner": "sourcecred",
          "type": "REPO",
        },
        "type": "ISSUE",
      },
      "type": "COMMENT",
    },
  },
  Object {
    "dst": Object {
      "id": "171460198",
      "parent": Object {
        "id": "100313899",
        "pull": Object {
          "number": "5",
          "repo": Object {
            "name": "example-github",
            "owner": "sourcecred",
            "type": "REPO",
          },
          "type": "PULL",
        },
        "type": "REVIEW",
      },
      "type": "COMMENT",
    },
    "src": Object {
      "id": "385576920",
      "parent": Object {
        "number": "2",
        "repo": Object {
          "name": "example-github",
          "owner": "sourcecred",
          "type": "REPO",
        },
        "type": "ISSUE",
      },
      "type": "COMMENT",
    },
  },
  Object {
    "dst": Object {
      "id": "100313899",
      "pull": Object {
        "number": "5",
        "repo": Object {
          "name": "example-github",
          "owner": "sourcecred",
          "type": "REPO",
        },
        "type": "PULL",
      },
      "type": "REVIEW",
    },
    "src": Object {
      "id": "385576920",
      "parent": Object {
        "number": "2",
        "repo": Object {
          "name": "example-github",
          "owner": "sourcecred",
          "type": "REPO",
        },
        "type": "ISSUE",
      },
      "type": "COMMENT",
    },
  },
  Object {
    "dst": Object {
      "number": "2",
      "repo": Object {
        "name": "example-github",
        "owner": "sourcecred",
        "type": "REPO",
      },
      "type": "ISSUE",
    },
    "src": Object {
      "id": "385223316",
      "parent": Object {
        "number": "6",
        "repo": Object {
          "name": "example-github",
          "owner": "sourcecred",
          "type": "REPO",
        },
        "type": "ISSUE",
      },
      "type": "COMMENT",
    },
  },
  Object {
    "dst": Object {
      "number": "2",
      "repo": Object {
        "name": "example-github",
        "owner": "sourcecred",
        "type": "REPO",
      },
      "type": "ISSUE",
    },
    "src": Object {
      "id": "369162222",
      "parent": Object {
        "number": "3",
        "repo": Object {
          "name": "example-github",
          "owner": "sourcecred",
          "type": "REPO",
        },
        "type": "PULL",
      },
      "type": "COMMENT",
    },
  },
]
`;
